### PR TITLE
#162948825 Write Tests for Working Functionality

### PR DESCRIPTION
**Pull request by MaggieChege**
**What Does this PR  do?**
- Write tests 
**Description of Tasks to be completed**
- Write tests for Login and Registration.
- Update Readme
- Update .gitignore

**How should this be manually tested?**

Run Tests

``` python manage.py test```

**What are the relevant pivotal tracker stories**

[#162948825](https://www.pivotaltracker.com/story/show/162948825)

[starts #162948825]

### CODE REVIEW

**Comment by andrewhingah** [comment]
I think the instructions on how to manually test need to be more descriptive. Say like from cloning or `git pull`, checking out to the relevant branch, etc.

**Comment by MaggieChege** [comment]
> I think the instructions on how to manually test need to be more descriptive. Say like from cloning or `git pull`, checking out to the relevant branch, etc.

@ignatiusukwuoma  should I add descriptive information that is already in the Readme?

File: authors/apps/authentication/test_user_login.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import json
+from django.urls import reverse
+from rest_framework.views import status
+from rest_framework.test import APITestCase,APIClient
+
+# from authors.apps.authentication.backends import JWTAuthentication
```
Comment (kwameasiago) [inline authors/apps/authentication/test_user_login.py]: It's not good practice to push commented code

Comment (MaggieChege) [reply authors/apps/authentication/test_user_login.py]: Okay let me get rid of that.

File: authors/apps/authentication/test_user_login.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            }
+        }
+        self.token={'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9'}
+
+    def test_login(self):
+        ''' if user is registered'''
+        register=self.client.post(self.login_url,
+            self.signup_data,
+            format='json')
+        self.
```
Comment (Georgeygigz) [inline authors/apps/authentication/test_user_login.py]: Kindly remove this

Comment (MaggieChege) [reply authors/apps/authentication/test_user_login.py]: OK, working on it.

**Comment by kwameasiago** [comment]
Try and stick to the engineering playbook conventions when making PRs
https://github.com/andela/engineering-playbook/tree/master/5.%20Developing/Conventions#branch-naming

**Review by andrewhingah — APPROVED** [review]
Looks good.

File: authors/apps/authentication/test_user_registration.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+from rest_framework.views import status
+from rest_framework.test import APITestCase,APIClient
+
+class RegistrationTestCase(APITestCase):
+    def setUp(self):
+        self.client = APIClient()
+        self.login_url =reverse('authentication:auth-login')
+        self.signup_url = reverse('authentication:auth-register')
+        self.signup_data = {
+  "user":{
```
Comment (Georgeygigz) [inline authors/apps/authentication/test_user_registration.py]: You have done good work, but I would recommend that you indent your code according to pep8 standards. You can check more on pep8 standards here https://www.python.org/dev/peps/pep-0008/#indentation.

File: authors/apps/authentication/test_user_registration.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            "email": "jacky@andela.com",
+            "password": "jakejake"
+            }}
+        self.duplicate_email2={
+  "user":{
+            "username":"jacky",
+            "email": "jacky@andela.com",
+            "password": "jakejake"
+            }}
+    def test_register_user(self):
```
Comment (Georgeygigz) [inline authors/apps/authentication/test_user_registration.py]: Have a consistent way of adding blank lines in your code for readability purposes. Read about blank lines on pep8 standards https://www.python.org/dev/peps/pep-0008/#indentation.

**Review by Georgeygigz — APPROVED** [review]
Looks good to me

**Comment by ignatiusukwuoma** [comment]
> > I think the instructions on how to manually test need to be more descriptive. Say like from cloning or `git pull`, checking out to the relevant branch, etc.
> 
> @ignatiusukwuoma should I add descriptive information that is already in the Readme?

No need, what you've done should be okay

File: authors/apps/authentication/test_user_login.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import json
+from django.urls import reverse
+from rest_framework.views import status
+from rest_framework.test import APITestCase,APIClient
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/test_user_login.py]: PEP 8 recommends whitespace after comma

File: authors/apps/authentication/test_user_registration.py (hunk 4 above)
Comment (ignatiusukwuoma) [reply authors/apps/authentication/test_user_registration.py]: You have not read and implemented this @MaggieChege

**Comment by ignatiusukwuoma** [comment]
Can we have the tests in a directory called tests within each application?

**Comment by ignatiusukwuoma** [comment]
Why are we only testing for the status code, why not testing the response object?

File: authors/apps/authentication/urls.py [hunk 6]
```diff
@@ -6,6 +6,6 @@
 app_name="authentication"
 urlpatterns = [
     path('user/', UserRetrieveUpdateAPIView.as_view()),
-    path('users/', RegistrationAPIView.as_view()),
-    path('users/login/', LoginAPIView.as_view()),
+    path('users/', RegistrationAPIView.as_view(),name="auth-register"),
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/urls.py]: Space after comma. Please read the PEP 8 standards on Indentation and implement throughout your code. You can search for a VSCode extension to help you with that.

**Review by kwameasiago — APPROVED** [review]
Looks good to me

**Comment by KMaina** [comment]
In your PR description, it is probably a good idea to include the PT story link and id. You can do it as follows -> 
[#Story ID](link to the story)
[#162948825](https://www.pivotaltracker.com/story/show/162948825)

**Comment by MaggieChege** [comment]
> In your PR description, it is probably a good idea to include the PT story link and id. You can do it as follows ->
> [#Story ID](link to the story)
> [#162948825](https://www.pivotaltracker.com/story/show/162948825)

Let me add it.

File: requirements.txt [hunk 7]
```diff
@@ -5,3 +5,4 @@ djangorestframework==3.9.0
 PyJWT==1.7.1
 six==1.12.0 
 psycopg2-binary==2.7.6.1
+pep8==1.7.1
```
Comment (ignatiusukwuoma) [inline requirements.txt]: pep8 has been renamed to `pycodestyle`.
Use of the pep8 tool will be removed in a future release.
Please install and use `pycodestyle` instead.

File: authors/apps/authentication/views.py [hunk 8]
```diff
@@ -72,4 +72,4 @@ def update(self, request, *args, **kwargs):
         serializer.save()
 
         return Response(serializer.data, status=status.HTTP_200_OK)
-
+
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/views.py]: Remove spaces on blank line

File: requirements.txt (hunk 7 above)
Comment (MaggieChege) [reply requirements.txt]: Okay I'll add that

File: authors/apps/authentication/views.py (hunk 8 above)
Comment (MaggieChege) [reply authors/apps/authentication/views.py]: Okay I've replaced  Pep8  with pycodestyle

File: requirements.txt [hunk 9]
```diff
@@ -5,3 +5,4 @@ djangorestframework==3.9.0
 PyJWT==1.7.1
 six==1.12.0 
 psycopg2-binary==2.7.6.1
+pycodestyle-2.4.0
```
Comment (ignatiusukwuoma) [inline requirements.txt]: Please include instructions for how to use pycodestyle in the README

Comment (MaggieChege) [reply requirements.txt]: Okay, I'll include that in the readme

File: authors/apps/authentication/tests/test_user_registration.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        self.assertEqual(response.content, {"user": {"email": "jake@jake.jake",
+                                            "username": "gigz"}})
+
+    def test_empty_username(self):
+        '''Test register user with empty username'''
+        response = self.client.post(self.signup_url,
+                                    self.signup_empty_username,
+                                    format="json")
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(json.loads(response.content), {"errors": {"username":
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/tests/test_user_registration.py:91]: why do we have to use json.loads here and not on line 82?

Comment (MaggieChege) [reply authors/apps/authentication/tests/test_user_registration.py:91]: They are two different tests the one on line 82 returns serialized data and the one on line 91  returns an error message.

File: authors/apps/authentication/tests/test_user_registration.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+                }}
+        self.duplicate_email2 = {
+            "user": {
+                "username": "jacky2",
+                "email": "jacky@andela.com",
+                "password": "jakejake"
+                }}
+
+    def test_register_user(self):
+        ''''Test register user'''
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/tests/test_user_registration.py]: There are four single-quoted strings here. Please use triple double-quoted strings instead for all docstrings.

Comment (MaggieChege) [reply authors/apps/authentication/tests/test_user_registration.py]: I missed that...thanks

File: authors/apps/authentication/tests/test_user_login.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+            "user": {
+                "email": "jake@jake.jake",
+                "password": "jakejake"
+                }}
+        self.login_unregistered_user_data = {
+            "user": {
+                "email": "jake@jake.jake",
+                "password": "jakejake"
+                }}
+        self.token = {'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9'}
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/tests/test_user_login.py]: what's the token for?

Comment (MaggieChege) [reply authors/apps/authentication/tests/test_user_login.py]: Its of no use at the moment, let me get rid of it.

**Review by ignatiusukwuoma — APPROVED** [review]

**Pull request merged**